nomos-sdp/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+        self.services
+            .get(&service_type)
+            .is_some_and(|service| service.contains(&provider_id))
+    }
+
+    pub fn insert_service_provider(&mut self, provider_id: ProviderId, service_type: ServiceType) {
+        self.services
+            .entry(service_type)
+            .or_default()
+            .insert(provider_id);
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct DeclarationUpdate {
+    pub declaration_id: DeclarationId,
+    pub provider_id: ProviderId,
+    pub service_type: ServiceType,
+    pub locators: Vec<Locator>,
+}
+
+impl<Proof> From<&DeclarationMessage<Proof>> for DeclarationUpdate {
+    fn from(message: &DeclarationMessage<Proof>) -> Self {
+        Self {
+            declaration_id: message.declaration_id(),
+            provider_id: message.provider_id,
+            service_type: message.service_type,
+            locators: message.locators.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct DeclarationMessage<Proof> {
+    pub service_type: ServiceType,
+    pub locators: Vec<Locator>,
+    pub proof_of_funds: Proof,
+    pub provider_id: ProviderId,
+}
+
+impl<Proof> DeclarationMessage<Proof> {
+    fn declaration_id(&self) -> DeclarationId {
+        let mut hasher = Blake2b::new();
+        for locator in &self.locators {
+            hasher.update(locator.addr.as_ref());
+        }
+        DeclarationId(hasher.finalize().into())
+    }
+}
+
+#[derive(Clone)]
+pub struct WithdrawMessage<Metadata> {
+    pub declaration_id: DeclarationId,
+    pub service_type: ServiceType,
+    pub provider_id: ProviderId,
+    pub nonce: Nonce,
+    pub metadata: Option<Metadata>,
+}
+
+#[derive(Clone)]
+pub struct RewardMessage<Metadata> {
+    pub declaration_id: DeclarationId,
+    pub service_type: ServiceType,
+    pub provider_id: ProviderId,
+    pub nonce: Nonce,
+    pub metadata: Option<Metadata>,
+}
+
+impl<Metadata> RewardMessage<Metadata> {
+    pub fn reward_id(&self) -> RewardId {
+        let mut hasher = Blake2b::new();
+        hasher.update(self.declaration_id.0);
+        hasher.update(self.provider_id.0);
+        hasher.update(self.nonce);
+        RewardId(hasher.finalize().into())
+    }
+}
+
+/// Withdrawal to Reward message conversion error
+///
+/// If withdrawal has no metadata, then it can't be converted to reward message
+/// and passed to the reward handling logic.
+pub struct NoMetadata;
+
+impl<Metadata> TryFrom<WithdrawMessage<Metadata>> for RewardMessage<Metadata> {
+    type Error = NoMetadata;
+
+    fn try_from(withdraw: WithdrawMessage<Metadata>) -> Result<Self, Self::Error> {
+        if withdraw.metadata.is_none() {
+            return Err(NoMetadata);
+        }
+
+        Ok(Self {
+            declaration_id: withdraw.declaration_id,
+            service_type: withdraw.service_type,
+            provider_id: withdraw.provider_id,
+            nonce: withdraw.nonce,
+            metadata: withdraw.metadata,
+        })
+    }
+}
```

Comment:
If we have a `NoMetadata` construct then it should not be optional in the struct message declaration. I'm not sure if the plan was to change it later on.

Reply:
> Reward message can have no metadata and still be rewarded. But - withdrawal message - if it has no metadata, then it can't be rewarded.

Yes, but for me it makes no sense that you can have a `Some(NoMetadata)` in any construct. Or I do not see the case at least.

Reply:
Hm, I'm puzzled, in what case we can have `Some(NoMetadata)`? This method (`TryFrom<Withdrawal> for Reward`) returns a result. So we'll either have `Ok(Reward)` or `Err(NoMetadata)`?